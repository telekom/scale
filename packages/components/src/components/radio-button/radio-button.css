/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

scale-radio-button {
  --color-disabled: var(--telekom-color-text-and-icon-disabled);
  --background-disabled: var(--telekom-color-ui-disabled);
  --color-error: var(--telekom-color-text-and-icon-functional-danger);
  --transition: all var(--telekom-motion-duration-transition)
    cubic-bezier(var(--telekom-motion-easing-standard));
  --color-primary: var(--telekom-color-text-and-icon-primary-standard);
  --color-primary-hover: var(--telekom-color-text-and-icon-standard);
  --color-primary-active: var(--telekom-color-text-and-icon-standard);
  --color-focus: var(--telekom-color-functional-focus);
  --color-text: var(--telekom-color-text-and-icon-standard);

  /*helper-text*/
  --transition-helper-text: var(--transition);
  --font-size-helper-text: var(--telekom-typography-font-size-small);
  --font-weight-helper-text: var(--telekom-typography-font-weight-bold);
  --line-height-helper-text: var(--telekom-typography-line-spacing-standard);
  --color-helper-text: var(
    --telekom-color-text-and-icon-functional-informational
  );
  --spacing-left-helper-text: var(--telekom-spacing-unit-x6);
  --spacing-top-helper-text: var(--telekom-spacing-unit-x1);

  /*label*/
  --transition-label: var(--transition);
  --font-weight-label: var(--telekom-typography-font-weight-medium);
  --color-label: var(--color-text);

  /*control*/
  --width-control: var(--telekom-spacing-unit-x4);
  --height-control: var(--telekom-spacing-unit-x4);
  --transition-control: var(--transition);
  --spacing-control: 0 var(--telekom-spacing-unit-x2) 0 0;
  --background-color-control: var(--telekom-color-ui-additional-base);
  --border-control: var(--telekom-spacing-unit-x025) solid
    var(--telekom-color-ui-outlined-border-standard);
  --border-control-checked: calc(0.5 * var(--telekom-spacing-unit-x3)) solid
    var(--telekom-color-primary-standard);
  --border-control-checked-active: calc(0.5 * var(--telekom-spacing-unit-x3))
    solid var(--telekom-color-primary-pressed, #cb0068);
  --border-control-checked-disabled: calc(0.5 * var(--telekom-spacing-unit-x3))
    solid var(--telekom-color-ui-outlined-border-disabled);
  --background-control-checked-disabled: var(
    --telekom-color-ui-outlined-fill-disabled
  );
  --border-control-active: var(--telekom-spacing-unit-x2) solid
    var(--telekom-color-primary-pressed);
  --border-control-disabled: var(--telekom-spacing-unit-x025) solid
    var(--telekom-color-ui-outlined-border-disabled);
  --border-control-error: var(--telekom-spacing-unit-x05) solid
    var(--telekom-color-functional-danger-standard);
  --box-shadow-control-focus: 0 0 0 var(--telekom-spacing-unit-x05)
    var(--color-focus);
}

.radio-button {
  position: relative;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  cursor: pointer;
  width: fit-content;
}

.radio-button .radio-button__meta {
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.radio-button .radio-button__helper-text {
  transition: var(--transition-helper-text);
  font-size: var(--font-size-helper-text);
  font-weight: var(--font-weight-helper-text);
  line-height: var(--line-height-helper-text);
  color: var(--color-helper-text);
  margin-top: var(--spacing-top-helper-text);
  padding-left: var(--spacing-left-helper-text);
}

.radio-button--status-error .radio-button__helper-text {
  color: var(--color-error);
}

.radio-button label {
  color: var(--color-label);
  transition: var(--transition-label);
  font-weight: var(--font-weight-label);
  cursor: pointer;
}
.radio-button input {
  width: var(--width-control);
  height: var(--height-control);
  transition: var(--transition-control);
  border-radius: 50%;
  -webkit-appearance: none;
  background-color: var(--telekom-color-ui-outlined-fill-standard);
  border: var(--border-control);
  margin: var(--spacing-control);
  cursor: pointer;
}
.radio-button input:focus {
  outline: none;
  box-shadow: var(--box-shadow-control-focus);
}
.radio-button:hover input:not(:checked):not([disabled]) {
  box-shadow: none;
  border-color: var(--telekom-color-ui-outlined-border-hovered);
  background-color: var(--telekom-color-ui-outlined-fill-hovered);
}
.radio-button:hover input:not(:checked):not([disabled]) ~ label {
  color: var(--color-primary-hover);
}
.radio-button:active input:not(:checked):not([disabled]) {
  border-color: var(--telekom-color-ui-outlined-border-pressed);
  background-color: var(--telekom-color-ui-outlined-fill-pressed);
}
.radio-button input:not(:checked):not([disabled]):active ~ label {
  color: var(--color-primary-active);
}
.radio-button input:disabled {
  border: var(--border-control-disabled);
}
.radio-button input:disabled ~ label {
  color: var(--color-disabled);
}
.radio-button.radio-button--status-error input {
  border: var(--border-control-error);
}
.radio-button.radio-button--status-error:hover
  input:not(:checked):not([disabled]) {
  border-color: var(--telekom-color-functional-danger-hovered);
}
.radio-button.radio-button--status-error:active
  input:not(:checked):not([disabled]) {
  border-color: var(--telekom-color-functional-danger-pressed);
}
.radio-button input:checked {
  border: var(--border-control-checked);
}
.radio-button input:checked:active {
  border: var(--border-control-checked-active);
}
.radio-button input:checked:disabled {
  background: var(--background-control-checked-disabled);
  border: var(--border-control-checked-disabled);
}
.radio-button input:checked:disabled ~ label {
  color: var(--color-disabled);
}

.radio-button--disabled label,
.radio-button--disabled .radio-button__label,
.radio-button--disabled input,
.radio-button--disabled .radio-button__helper-text {
  cursor: not-allowed;
  border-color: var(--color-disabled);
  color: var(--color-disabled);
  background: var(--telekom-color-ui-outlined-fill-standard);
}
