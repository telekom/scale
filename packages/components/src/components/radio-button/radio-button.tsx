/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, Event, EventEmitter, h, Host, Prop } from '@stencil/core';
import classNames from 'classnames';

interface InputChangeEventDetail {
  value: string | number | boolean | undefined | null;
}

let i = 0;

@Component({
  tag: 'scale-radio-button',
  styleUrl: './radio-button.css',
  shadow: false,
})
export class RadioButton {
  /** (optional) Input name */
  @Prop() name?: string = '';
  /** (optional) Input label */
  @Prop() label: string = '';
  /** (optional) Input helper text */
  @Prop() helperText?: string = '';
  /** (optional) Input status */
  @Prop() status?: string = '';
  /** (optional) Input disabled */
  @Prop() disabled?: boolean;
  /** (optional) Input checked */
  @Prop({ reflect: true }) checked?: boolean = false;
  /** (optional) Input value */
  @Prop({ mutable: true }) value?: string | number | null = '';
  /** (optional) Input checkbox id */
  @Prop() inputId?: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  @Event() scaleChange!: EventEmitter<InputChangeEventDetail>;

  componentWillLoad() {
    if (this.inputId == null) {
      this.inputId = 'input-' + i++;
    }
  }

  handleCheckedChange = (event: any) => {
    this.checked = event.target.checked;
    // I don't think this is ever going to be `false` but well...
    if (this.checked) {
      this.uncheckSiblings();
    }
    this.scaleChange.emit({
      value: this.value == null ? this.value : this.value.toString(),
    });
  };

  // Prevent click event being fired twice when the target is the label.
  handleLabelClick = (event: any) => {
    event.stopPropagation();
  };

  // We manually set `checked` to false on sibling <scale-radio-button> elements,
  // otherwise they stayed `checked` after being clicked once, forever.
  uncheckSiblings() {
    this.getSiblingRadios().forEach((radio: HTMLScaleRadioButtonElement) => {
      radio.checked = false;
    });
  }

  getSiblingRadios(): HTMLElement[] {
    return Array.from(
      document.querySelectorAll(`scale-radio-button[name="${this.name}"]`)
    ).filter(
      (radio: HTMLScaleRadioButtonElement) => radio.inputId !== this.inputId
    ) as HTMLScaleRadioButtonElement[];
  }

  render() {
    const ariaInvalidAttr =
      this.status === 'error' ? { 'aria-invalid': true } : {};
    const helperTextId = `helper-message-${i}`;
    const ariaDescribedByAttr = { 'aria-describedBy': helperTextId };

    return (
      <Host>
        <div class={this.getCssClassMap()}>
          <input
            type="radio"
            name={this.name}
            id={this.inputId}
            onChange={this.handleCheckedChange}
            value={this.value}
            checked={this.checked}
            disabled={this.disabled}
            {...ariaInvalidAttr}
            {...(this.helperText ? ariaDescribedByAttr : {})}
          />
          <label htmlFor={this.inputId} onClick={this.handleLabelClick}>
            {this.label}
          </label>
          {!!this.helperText && (
            <div
              class="radio-button__meta"
              id={helperTextId}
              aria-live="polite"
              aria-relevant="additions removals"
            >
              <div class="radio-button__helper-text">{this.helperText}</div>
            </div>
          )}
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames(
      'radio-button',
      this.checked && `radio-button--checked`,
      this.disabled && `radio-button--disabled`,
      this.status && `radio-button--status-${this.status}`
    );
  }
}
