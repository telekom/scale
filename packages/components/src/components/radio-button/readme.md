# scale-radio-button



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description                    | Type               | Default     |
| ------------ | ------------- | ------------------------------ | ------------------ | ----------- |
| `checked`    | `checked`     | (optional) Input checked       | `boolean`          | `false`     |
| `disabled`   | `disabled`    | (optional) Input disabled      | `boolean`          | `undefined` |
| `helperText` | `helper-text` | (optional) Input helper text   | `string`           | `''`        |
| `inputId`    | `input-id`    | (optional) Input checkbox id   | `string`           | `undefined` |
| `label`      | `label`       | (optional) Input label         | `string`           | `''`        |
| `name`       | `name`        | (optional) Input name          | `string`           | `''`        |
| `status`     | `status`      | (optional) Input status        | `string`           | `''`        |
| `styles`     | `styles`      | (optional) Injected CSS styles | `string`           | `undefined` |
| `value`      | `value`       | (optional) Input value         | `number \| string` | `''`        |


## Events

| Event          | Description                                                                                        | Type                                  |
| -------------- | -------------------------------------------------------------------------------------------------- | ------------------------------------- |
| `scale-change` |                                                                                                    | `CustomEvent<InputChangeEventDetail>` |
| `scaleChange`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<InputChangeEventDetail>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
