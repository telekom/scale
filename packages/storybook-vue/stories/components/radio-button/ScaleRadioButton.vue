<template>
  <scale-radio-button
    :name="name"
    :invalid="invalid"
    :label="label"
    :helper-text="helperText"
    :disabled="disabled"
    :checked="checked"
    :value="value"
    :input-id="inputId"
    @scaleChange="scaleChange"
  >
    <slot></slot>
  </scale-radio-button>
</template>

<script>
import { action } from "@storybook/addon-actions";

export default {
  props: {
    invalid: { type: Boolean },
    name: { type: String },
    label: { type: String },
    helperText: { type: String },
    disabled: { type: Boolean },
    checked: { type: Boolean },
    value: { type: String },
    inputId: { type: String },
    styles: { type: String }
  },
  methods: {
    scaleChange($event) {
      action("scaleChange");
      this.$emit("scaleChange", $event);
    },
    'scale-change'($event) {
      action("scale-change");
      this.$emit("scale-change", $event);
    },
  }
};
</script>
