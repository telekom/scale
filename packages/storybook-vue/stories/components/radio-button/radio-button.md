<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Radio Button</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

Radio buttons allow a user to only select one option from a collection of choices.

## General

A radio button is a selection control that consists of a circular button and a label. It is considered to be on with a filled circle and off with an empty circle.

Radio buttons only appear in groups. Activating a radio button automatically deselects the previously selected option.

![Image Name](assets/3_components/radio-button/radiobuttons_general.png)

### When to use

Use radio buttons to provide the user with a collection of related but mutually exclusive choices.

### When not to use

If the user needs to choose one or more options, use checkboxes.
To initiate an action, use a button.

## Elements

![Image Name](assets/3_components/radio-button/radio-button-elements.png)

### Radio button deselected (1)

An empty circle is considered to be deselected.

### Radio button selected (2)

If the circle is filled, the option is considered to be selected.

### Information (3) (Optional)

When necessary, you can also provide additional information to help users.

### Error (4)

If the selection is contradictory or not applicable, an error message appears.

## Label

The label clearly describes the effect of activating the radio button. Labels normally contain at most three words.

![Image Name](assets/3_components/radio-button/radiobuttons_label.png)

## Disabled state

Disable the radio button if:

- another action has to take place before the radio button is usable or
- the option needs to be listed but is currently unavailable.

![Image Name](assets/3_components/radio-button/disabled_radiobutton.png)

> The disabled state is exempt from the WCAG contrast minimum for text colors. You can find more information in the [WCAG guidelines](https://www.w3.org/TR/WCAG21/#contrast-minimum).

## Headline and default selection

Consider using a headline to provide more clarity to the user.

Display one option as a default selection to align with user expectations and reduce the time and clicks needed. This is also a good way of making suggestions to the user. There are a few cases where you don't have to select a radio button by default, e.g., when selecting titles in a form (Mr., Ms., etc.).

![Image Name](assets/3_components/radio-button/radiobuttons_bestpractice.png)

## Related components

[Checkbox, ](?path=/usage/components-checkbox--standard)
[Switch, ](?path=/usage/components-switch--standard)
[Dropdown ](?path=/usage/components-dropdown--standard)
