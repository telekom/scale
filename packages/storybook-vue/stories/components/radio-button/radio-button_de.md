<div style="display: inline-flex; align-items: center; justify-content: space-between; width: 100%;">
    <h1>Radio Button</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

Radio-Buttons ermöglichen es Nutzer\*innen, eine Option aus einer Reihe von Alternativen auszuwählen.

## Allgemein

Ein Radio-Button ist ein Steuerelement und besteht aus einem runden Button mit einem Label. Die Nutzer\*innen können per Klick/Tap eine Option auswählen (Kreis ist gefüllt) oder wieder abwählen (Kreis ist leer).

Radio-Buttons funktionieren nur als Gruppe. Durch Aktivieren eines Radio-Buttons wird die zuvor gewählte Option automatisch abgewählt.

![Image Name](assets/3_components/radio-button/radiobuttons_general.png)

### Verwende Radio-Buttons, um

- Nutzer\*innen eine Reihe von verwandten, aber sich gegenseitig ausschließenden Auswahlmöglichkeiten anzubieten.

### Verwende Radio-Buttons besser nicht, wenn

- Nutzer\*innen mehr als eine Option auswählen sollen (nutze hierfür Checkboxen) oder
- eine Aktion ausführen müssen (nutze hierfür einen Button).

## Elemente

![Image Name](assets/3_components/radio-button/radio-button-elements.png)

### Radio-Button ausgeschaltet (1)

Wenn der Kreis leer ist, gilt eine Option als abgewählt.

### Radio-Button eingeschaltet (2)

Ist der Kreis ausgefüllt, so gilt eine Option als ausgewählt.

### Information (3) (optional)

Wenn nötig, kannst du auch noch eine zusätzliche Information angeben.

### Fehlermeldung (4)

Bei widersprüchlicher oder nichtzutreffender Auswahl erscheint eine Fehlermeldung.

## Label

Beschreibe mit dem Label deutlich, welche Auswirkung die jeweilige Auswahl hat. Ein Label enthält in der Regel maximal drei Wörter.

![Image Name](assets/3_components/radio-button/radiobuttons_label.png)

## Deaktivierter Zustand

Deaktiviere den Radio-Button, wenn

- eine andere Aktion erfolgen muss, bevor der Radio-Button nutzbar ist oder
- die Option aufgeführt werden muss, aber aktuell nicht auswählbar ist.

![Image Name](assets/3_components/radio-button/disabled_radiobutton.png)

## Titel und Default-Auswahl

Treffe eine Vorauswahl durch die Default-Aktivierung einer Option. Entspricht diese den Erwartungen der Nutzer\*innen, reduzierst du somit den benötigen Interaktionsaufwand. Das ist zudem eine gute Möglichkeit, Nutzer\*innen Vorschläge zu machen.

Es gibt ein paar wenige Fälle, in denen du besser keine Vorauswahl triffst, zum Beispiel bei der Auswahl des Titels in einem Formular (Herr, Frau etc.).

In vielen Fällen ist es für Nutzer\*innen hilfreich, wenn du eine Headline verwendest.

![Image Name](assets/3_components/radio-button/radiobuttons_bestpractice.png)

## Verwandte Komponenten

<a href="?path=/usage/components-checkbox--standard">Checkbox, </a>
<a href="?path=/usage/components-switch--standard">Switch, </a>
<a href="?path=/usage/components-dropdown--standard">Dropdown</a>
