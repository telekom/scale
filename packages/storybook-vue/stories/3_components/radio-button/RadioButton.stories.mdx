import { Meta, Story, ArgsTable, Canvas, Description } from "@storybook/addon-docs/blocks";
import ScaleRadioButton from "./ScaleRadioButton.vue";
import { action } from "@storybook/addon-actions";
export const ICON = "M16.471,8.321 L12.613594,0.502886112 L8.755,8.321 L0.126435258,9.57533801 L6.37,15.66 L4.89610548,24.2548735 L12.613,20.197 L20.3310826,24.2548735 L18.856,15.66 L25.1007528,9.57533801 L16.471,8.321 Z M9.41891527,9.23556376 L12.613,2.761 L15.8082728,9.23556376 L22.951,10.273 L17.7826928,15.3122039 L19.003,22.427 L12.613594,19.067774 L6.223,22.427 L7.44449521,15.3122039 L2.275,10.273 L9.41891527,9.23556376 Z";

<Meta
  title="Components/Radio Button"
  component={ScaleRadioButton}
  argTypes={{}}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleRadioButton },
  props: {
    label: String,
    ...ScaleRadioButton.props
  },
  template: `
    <scale-radio-button
      :status="status"
      :name="name"
      :label="label"
      :helper-text="helperText"
      :disabled="disabled"
      :checked="checked"
      :value="value"
      :input-id="inputId"
      @scaleChange="scaleChange"
    ></scale-radio-button>
  `,
  methods: {
    scaleChange: action("scaleChange"),
  }
})

<div style={{display: 'inline-flex', alignItems: 'center', justifyContent: 'space-between', width: '100%'}}>
    <h1>Radio Button</h1>
    <img src="assets/aa.png" alt="Accessible AA" />
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: "Radio"
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

### Scoped CSS variables

```css
scale-radio-button {
  --color-disabled: var(--scl-color-background-disabled);
  --background-disabled: var(--scl-color-white);
  --color-error: var(--scl-color-text-error);
  --transition: all var(--scl-motion-duration-fast)
    var(--scl-motion-easing-standard);
  --color-primary: var(--scl-color-primary, #e20074);
  --color-primary-hover: var(--scl-color-primary-hover, #f90984);
  --color-primary-active: var(--scl-color-primary-active);
  --color-focus: var(--scl-color-focus);
  --color-text: var(--scl-color-text-standard);

  /*helper-text*/
  --transition-helper-text: var(--transition);
  --font-size-helper-text: var(--scl-font-size-12);
  --font-weight-helper-text: var(--scl-font-weight-bold);
  --line-height-helper-text: var(--scl-font-line-height-133);
  --color-helper-text: var(--scl-color-blue-70);
  --spacing-left-helper-text: var(--scl-spacing-24);
  --spacing-top-helper-text: var(--scl-spacing-4);

  /*label*/
  --transition-label: var(--transition);
  --font-weight-label: var(--scl-font-weight-medium);
  --color-label: var(--color-text);

  /*control*/
  --width-control: var(--scl-spacing-16);
  --height-control: var(--scl-spacing-16);
  --transition-control: var(--transition);
  --spacing-control: 0 var(--scl-spacing-8) 0 0;
  --background-color-control: var(--scl-color-white);
  --border-control: var(--scl-spacing-1) solid var(--color-text);
  --border-control-checked: calc(0.5 * var(--scl-spacing-12)) solid
    var(--color-primary);
  --border-control-checked-active: calc(0.5 * var(--scl-spacing-12)) solid
    var(--color-primary-active, #cb0068);
  --border-control-checked-disabled: calc(0.5 * var(--scl-spacing-12)) solid
    var(--color-disabled);
  --background-control-checked-disabled: var(--scl-color-grey-30);
  --border-control-active: var(--scl-spacing-8) solid
    var(--color-primary-active);
  --border-control-disabled: var(--scl-spacing-1) solid var(--color-disabled);
  --border-control-error: var(--scl-spacing-2) solid
    var(--scl-color-background-error);
  --box-shadow-control-focus: 0 0 0 var(--scl-spacing-2) var(--color-focus);
}
```

## Standard Disabled

<Canvas withSource="none">
  <Story
    name="Standard Disabled"
    args={{
      label: "Standard Disabled",
      disabled: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button disabled></scale-radio-button>
```


## Selected

<Canvas withSource="none">
  <Story
    name="Selected"
    args={{
      label: "Selected",
      checked: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button checked></scale-radio-button>
```


## Selected Disabled

<Canvas withSource="none">
  <Story
    name="Selected Disabled"
    args={{
      label: "Selected Disabled",
      checked: true,
      disabled: true
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button checked disabled></scale-radio-button>
```

## Helper Text

<Canvas withSource="none">
    <Story
        name="Helper Text"
        args={{
            label: "Helper Text",
            helperText: "Make sure to select this"
        }}
    >
        {Template.bind({})}
    </Story>
</Canvas>

```html
<scale-radio-button label="Helper Text" helper-text="Make sure to select this"></scale-radio-button>
```

## Error

<Canvas withSource="none">
  <Story
    name="Error"
    args={{
      label: "Standard with error",
      status: "error",
      helperText: "This is the error message"
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button status="error" helper-text="This is the error message"></scale-radio-button>
```
