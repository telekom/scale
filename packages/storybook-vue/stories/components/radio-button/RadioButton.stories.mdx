import {
  Meta,
  Story,
  ArgsTable,
  Canvas,
  Description,
} from '@storybook/addon-docs';
import ScaleRadioButton from './ScaleRadioButton.vue';
import { action } from '@storybook/addon-actions';
export const ICON =
  'M16.471,8.321 L12.613594,0.502886112 L8.755,8.321 L0.126435258,9.57533801 L6.37,15.66 L4.89610548,24.2548735 L12.613,20.197 L20.3310826,24.2548735 L18.856,15.66 L25.1007528,9.57533801 L16.471,8.321 Z M9.41891527,9.23556376 L12.613,2.761 L15.8082728,9.23556376 L22.951,10.273 L17.7826928,15.3122039 L19.003,22.427 L12.613594,19.067774 L6.223,22.427 L7.44449521,15.3122039 L2.275,10.273 L9.41891527,9.23556376 Z';

<Meta
  title="Components/Radio Button"
  component={ScaleRadioButton}
  argTypes={{}}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleRadioButton },
  props: {
    label: String,
    ...ScaleRadioButton.props,
  },
  template: `
    <scale-radio-button
      :invalid="invalid"
      :name="name"
      :label="label"
      :helper-text="helperText"
      :disabled="disabled"
      :checked="checked"
      :value="value"
      :input-id="inputId"
      @scaleChange="scaleChange"
    ></scale-radio-button>
  `,
  methods: {
    scaleChange: action('scaleChange'),
  },
});

<div
  style={{
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    width: '100%',
  }}
>
  <h1>Radio Button</h1>
  <img src="assets/aa.png" alt="Accessible AA" />
</div>

## Standard

<Canvas withSource="none">
  <Story
    name="Standard"
    args={{
      label: 'Radio',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<ArgsTable story="Standard" />

### Scoped CSS variables

```css
scale-radio-button {
  --color-disabled: var(--telekom-color-text-and-icon-disabled);
  --background-disabled: var(--telekom-color-ui-disabled);
  --color-error: var(--telekom-color-text-and-icon-functional-danger);
  --transition: all var(--telekom-motion-duration-transition) var(
      --telekom-motion-easing-standard
    );
  --color-primary: var(--telekom-color-text-and-icon-primary-standard);
  --color-primary-hover: var(--telekom-color-text-and-icon-standard);
  --color-primary-active: var(--telekom-color-text-and-icon-standard);
  --color-focus: var(--telekom-color-functional-focus);
  --color-text: var(--telekom-color-text-and-icon-standard);

  /*helper-text*/
  --transition-helper-text: var(--transition);
  --font-size-helper-text: var(--telekom-typography-font-size-small);
  --font-weight-helper-text: var(--telekom-typography-font-weight-bold);
  --line-height-helper-text: var(--telekom-typography-line-spacing-standard);
  --color-helper-text: var(
    --telekom-color-text-and-icon-functional-informational
  );
  --spacing-left-helper-text: var(--telekom-spacing-unit-x6);
  --spacing-top-helper-text: var(--telekom-spacing-unit-x1);

  /*label*/
  --transition-label: var(--transition);
  --font-weight-label: var(--telekom-typography-font-weight-medium);
  --color-label: var(--color-text);

  /*control*/
  --width-control: var(--telekom-spacing-unit-x4);
  --height-control: var(--telekom-spacing-unit-x4);
  --transition-control: var(--transition);
  --spacing-control: 0 var(--telekom-spacing-unit-x2) 0 0;
  --background-color-control: var(--telekom-color-ui-additional-base);
  --border-control: var(--telekom-spacing-unit-x025) solid var(
      --telekom-color-ui-outlined-border-standard
    );
  --border-control-checked: calc(0.5 * var(--telekom-spacing-unit-x3)) solid var(
      --telekom-color-primary-standard
    );
  --border-control-checked-active: calc(0.5 * var(--telekom-spacing-unit-x3)) solid
    var(--telekom-color-primary-pressed, #cb0068);
  --border-control-checked-disabled: calc(0.5 * var(--telekom-spacing-unit-x3)) solid
    var(--telekom-color-ui-outlined-border-disabled);
  --background-control-checked-disabled: var(
    --telekom-color-ui-outlined-fill-disabled
  );
  --border-control-active: var(--telekom-spacing-unit-x2) solid var(
      --telekom-color-primary-pressed
    );
  --border-control-disabled: var(--telekom-spacing-unit-x025) solid var(
      --telekom-color-ui-outlined-border-disabled
    );
  --border-control-error: var(--telekom-spacing-unit-x05) solid var(
      --telekom-color-functional-danger-standard
    );
  --box-shadow-control-focus: 0 0 0 var(--telekom-spacing-unit-x05) var(
      --color-focus
    );
}
```

## Standard Disabled

<Canvas withSource="none">
  <Story
    name="Standard Disabled"
    args={{
      label: 'Standard Disabled',
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button disabled></scale-radio-button>
```

## Selected

<Canvas withSource="none">
  <Story
    name="Selected"
    args={{
      label: 'Selected',
      checked: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button checked></scale-radio-button>
```

## Selected Disabled

<Canvas withSource="none">
  <Story
    name="Selected Disabled"
    args={{
      label: 'Selected Disabled',
      checked: true,
      disabled: true,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button checked disabled></scale-radio-button>
```

## Helper Text

<Canvas withSource="none">
  <Story
    name="Helper Text"
    args={{
      label: 'Helper Text',
      helperText: 'Make sure to select this',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button
  label="Helper Text"
  helper-text="Make sure to select this"
></scale-radio-button>
```

## Error

<Canvas withSource="none">
  <Story
    name="Error"
    args={{
      label: 'Standard with error',
      invalid: true,
      helperText: 'This is the error message',
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

```html
<scale-radio-button
  invalid
  helper-text="This is the error message"
></scale-radio-button>
```
