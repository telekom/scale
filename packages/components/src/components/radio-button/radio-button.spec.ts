/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { RadioButton } from './radio-button';

describe('RadioButton', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [RadioButton],
      html: `<scale-radio-button></scale-radio-button>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should handle change of checked', async () => {
    const page = await newSpecPage({
      components: [RadioButton],
      html: `<scale-radio-button inputId="testing" checked="false"></scale-radio-button>`,
    });
    page.rootInstance.checked = true;
    expect(page.root).toMatchSnapshot();
  });

  it('should emit on change', async () => {
    const page = await newSpecPage({
      components: [RadioButton],
      html: `<scale-radio-button></scale-radio-button>`,
    });
    const changeSpy = jest.fn();
    const changeSpyLegacy = jest.fn();
    page.rootInstance.checked = true;
    page.doc.addEventListener('scale-change', changeSpy);
    page.doc.addEventListener('scaleChange', changeSpyLegacy);
    const element = page.root.querySelector('input');
    element.dispatchEvent(new Event('change'));
    await page.waitForChanges();
    expect(changeSpy).toHaveBeenCalled();
    expect(changeSpyLegacy).toHaveBeenCalled();
  });
});
