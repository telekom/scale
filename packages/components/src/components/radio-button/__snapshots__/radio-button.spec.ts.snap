// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton should handle change of checked 1`] = `
<scale-radio-button checked="false" inputid="testing">
  <div class="radio-button">
    <input id="input-1" type="radio" value="">
    <label htmlfor="input-1"></label>
  </div>
</scale-radio-button>
`;

exports[`RadioButton should match snapshot 1`] = `
<scale-radio-button>
  <div class="radio-button">
    <input id="input-0" type="radio" value="">
    <label htmlfor="input-0"></label>
  </div>
</scale-radio-button>
`;
